pkg/provision/storage/perWorkspaceStorage.go

```diff
@@ -0,0 +1,197 @@
+func (p *PerWorkspaceStorageProvisioner) rewriteContainerVolumeMounts(workspaceId, pvcName string, podAdditions *v1alpha1.PodAdditions, workspace *dw.DevWorkspaceTemplateSpec) error {
+	devfileVolumes := map[string]dw.VolumeComponent{}
+
+	// Construct map of volume name -> volume Component
+	for _, component := range workspace.Components {
+		if component.Volume != nil {
+			if _, exists := devfileVolumes[component.Name]; exists {
+				return fmt.Errorf("volume component '%s' is defined multiple times", component.Name)
+			}
+			devfileVolumes[component.Name] = *component.Volume
+		}
+	}
+
+	// Containers in podAdditions may reference e.g. automounted volumes in their volumeMounts, and this is not an error
+	additionalVolumes := map[string]bool{}
+	for _, additionalVolume := range podAdditions.Volumes {
+		additionalVolumes[additionalVolume.Name] = true
+	}
+
+	// Add implicit projects volume to support mountSources, if needed
+	if _, exists := devfileVolumes[devfileConstants.ProjectsVolumeName]; !exists {
+		projectsVolume := dw.VolumeComponent{}
+		projectsVolume.Size = constants.PVCStorageSize
+		devfileVolumes[devfileConstants.ProjectsVolumeName] = projectsVolume
+	}
+
+	// TODO: What should we do when a volume isn't explicitly defined?
+	rewriteVolumeMounts := func(containers []corev1.Container) error {
+		for cIdx, container := range containers {
+			for vmIdx, vm := range container.VolumeMounts {
+				volume, ok := devfileVolumes[vm.Name]
+				if !ok {
+					// Volume is defined outside of the devfile
+					if additionalVolumes[vm.Name] {
+						continue
+					}
+					// Should never happen as flattened Devfile is validated.
+					return fmt.Errorf("container '%s' references undefined volume '%s'", container.Name, vm.Name)
+				}
+				if !isEphemeral(&volume) {
+					containers[cIdx].VolumeMounts[vmIdx].SubPath = vm.Name
+					containers[cIdx].VolumeMounts[vmIdx].Name = pvcName
+				}
+			}
+		}
+		return nil
+	}
+	if err := rewriteVolumeMounts(podAdditions.Containers); err != nil {
+		return err
+	}
+	if err := rewriteVolumeMounts(podAdditions.InitContainers); err != nil {
+		return err
+	}
+
+	podAdditions.Volumes = append(podAdditions.Volumes, corev1.Volume{
+		Name: pvcName,
+		VolumeSource: corev1.VolumeSource{
+			PersistentVolumeClaim: &corev1.PersistentVolumeClaimVolumeSource{
+				ClaimName: pvcName,
+			},
+		},
+	})
+
+	return nil
+}
+
+func syncPerWorkspacePVC(workspace *dw.DevWorkspace, clusterAPI sync.ClusterAPI) (*corev1.PersistentVolumeClaim, error) {
+	namespace := workspace.Namespace
+
+	namespacedConfig, err := nsconfig.ReadNamespacedConfig(namespace, clusterAPI)
+	if err != nil {
+		return nil, fmt.Errorf("failed to read namespace-specific configuration: %w", err)
+	}
+
+	// TODO: Determine the storage size that is needed by iterating through workspace volumes,
+	// adding the sizes specified and figuring out overrides/defaults
```

Comment:
This TODO is planned to be addressed as part of https://github.com/devfile/devworkspace-operator/issues/740, right?

Reply:
Correct, or at the least, it might end up being a separate but related issue